### ISSUE TITLE
InstanceConfiguration extra loaders not thread safe

### ISSUE THREAD

**Issue by ccutrer** [jruby/jruby#7338]
when working in a highly multithreaded environment, that's adding additional class loaders (like an OSGi app that we're registering each bundle with JRuby as it loads up so it can be accessed from ruby), a concurrent modification error can become highly likely when one thread is trying to load a java class into jruby, and another is adding a new loader. Stack trace:

```
#<LoadError: load error: rspec/openhab/core/mocks/event_admin -- java.util.ConcurrentModificationException: null>
org/jruby/RubyKernel.java:1017:in `require'
/home/cody/src/rspec-openhab-scripting/lib/rspec/openhab/karaf.rb:212:in `block in launch_karaf'
/home/cody/src/rspec-openhab-scripting/lib/rspec/openhab/karaf.rb:343:in `block in set_up_service_listener'
org/jruby/java/proxies/ArrayJavaProxy.java:455:in `each'
/home/cody/src/rspec-openhab-scripting/lib/rspec/openhab/karaf.rb:339:in `block in set_up_service_listener'
cause: #<Java::JavaUtil::ConcurrentModificationException: >
java.base/java.util.LinkedList$ListItr.checkForComodification(LinkedList.java:970)
java.base/java.util.LinkedList$ListItr.next(LinkedList.java:892)
org.jruby.dist/org.jruby.javasupport.JavaSupport.loadJavaClass(JavaSupport.java:101)
org.jruby.dist/org.jruby.javasupport.Java.loadJavaClass(Java.java:1051)
org.jruby.dist/org.jruby.javasupport.Java.loadJavaClass(Java.java:1040)
org.jruby.dist/org.jruby.javasupport.Java.getProxyClassOrNull(Java.java:1091)
org.jruby.dist/org.jruby.javasupport.Java.getProxyOrPackageUnderPackage(Java.java:970)
org.jruby.dist/org.jruby.javasupport.JavaPackage.method_missing(JavaPackage.java:253)
org.jruby.dist/org.jruby.javasupport.JavaPackage$INVOKER$i$method_missing.call(JavaPackage$INVOKER$i$method_missing.gen)
<snip>
org.jruby.dist/org.jruby.Ruby.runInterpreter(Ruby.java:1272)
org.jruby.dist/org.jruby.Ruby.loadFile(Ruby.java:2867)
org.jruby.dist/org.jruby.runtime.load.LibrarySearcher$ResourceLibrary.load(LibrarySearcher.java:937)
org.jruby.dist/org.jruby.runtime.load.LibrarySearcher$FoundLibrary.load(LibrarySearcher.java:895)
org.jruby.dist/org.jruby.runtime.load.LoadService.tryLoadingLibraryOrScript(LoadService.java:665)
org.jruby.dist/org.jruby.runtime.load.LoadService.lambda$smartLoadInternal$0(LoadService.java:526)
org.jruby.dist/org.jruby.runtime.load.LoadService$RequireLocks.executeAndClearLock(LoadService.java:459)
org.jruby.dist/org.jruby.runtime.load.LoadService$RequireLocks.lock(LoadService.java:423)
org.jruby.dist/org.jruby.runtime.load.LoadService$RequireLocks.access$300(LoadService.java:378)
org.jruby.dist/org.jruby.runtime.load.LoadService.smartLoadInternal(LoadService.java:512)
org.jruby.dist/org.jruby.runtime.load.LoadService.require(LoadService.java:363)
org.jruby.dist/org.jruby.RubyKernel.requireCommon(RubyKernel.java:1024)
org.jruby.dist/org.jruby.RubyKernel.require(RubyKernel.java:1017)
<snip>
```

Workaround is simple from my code - I just wrapped a mutex around my addLoader call, and also my `require` that's referencing "new" Java classes. In JRuby, it might be easier to just make a copy of extraLoaders before iterating it.

**Environment Information**

 * `jruby 9.3.8.0-SNAPSHOT (2.6.8) 2022-08-29 02f5e6b04b OpenJDK 64-Bit Server VM 11.0.16+8-post-Ubuntu-0ubuntu122.04 on 11.0.16+8-post-Ubuntu-0ubuntu122.04 [x86_64-linux]` (local build with one additional commit: https://github.com/jruby/jruby/commits/d7a87d41c7c48cc311d15c92c3cad2fbe56834f5)  
Provide at least:
 * `Linux opennuc 5.15.0-46-generic #49-Ubuntu SMP Thu Aug 4 18:03:25 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux`

**Comment by kares** [comment]
interesting use-case, should be easy to switch to a concurrent-friendly list e.g. `CopyOnWriteArrayList`

but are you sure things would operate correctly - doesn't adding additional loaders mean to always happen before starting loading classes from the JRuby runtime?

**Comment by ccutrer** [comment]
it's no different than `require` needing to be thread safe in vanilla Ruby. and if you have a multi-threaded environment, your code just needs to be careful to not reference any constants defined by a require from another thread until you're sure they've been fully required.

in my case, I'm having JRuby be the main entry point in the process, and then it's booting up Karaf and OSGi. OSGi has callbacks that let me know a new bundle has been loaded, at which point I add the bundles I need to JRuby's extra loaders, and as appropriate require additional ruby code that extends/depends on the OSGi bundle just loaded. So my code itself is thread safe. But OSGi is doing lots of stuff in different threads, so a completely different thread might notify that another bundle was loaded, and I add it to JRuby's loaders, while the first thread is requiring my ruby code and looking up a java class, and 💥. Honestly I'm surprised at how reliable it is to reproduce for me.

**Comment by headius** [comment]
I'll flip these to use `CopyOnWriteArrayList` since these are very rarely written two and shouldn't be very long in general.

**Comment by headius** [comment]
Should be fixed on 9.3 branch by #7345.
